### ISSUE TITLE
Gives memory error with number of annotations greater than 200k

### ISSUE THREAD

**Issue by tanishqackerman** [MiXaiLL76/faster_coco_eval#57]
When i ran the .evaluate() function with dataset consisting 220k annotations it gave memory error.

**Comment by jveitchmichaelis** [comment]
I'm happy to provide a real test annotation dataset for 2M (ish?) annotations / 20k images. I assume you don't need the underlying imagery, which would be infeasible to transfer. Would COCO JSON be the best format?

I've been able to run this on a node with ~96 GB of RAM, as we had OOMs with 64 GB but I'm not sure what exactly is required by the COCO API and how much is the training pipeline/dataloader. The main concern we have is that on a large machine, resource utilization is poor since the metric only uses a single thread to compute. So if we could have an option parallelize over images, that would make a lot of sense (annotation matching is the bottleneck?).

**Comment by jveitchmichaelis** [comment]
No problem, let me check what I can share, but it should be fine. I will also write up some minimal code to profile the timing. Like I said, we didn't run into an OOM running the compute stage, but we use a big node.

The working memory would mostly depend on how you compute the cost matrix to assign pred x gt boxes for each image, and what has to be stored before you reduce over images. Once you've computed the gt <> pred IDs I think you can throw the cost information away and what's left is a memory allocation that is O(annotations) with the IoU, scores, etc?

**Comment by MiXaiLL76** [comment]
@jveitchmichaelis You haven't given me the data yet, but I started reworking the dataset (C++) a while ago, and just now I've done a proof-of-concept.

https://github.com/MiXaiLL76/faster_coco_eval/tree/fix-memory-bug

You can check it out with your tests, or I'll wait for your dataset.

**Comment by jveitchmichaelis** [comment]
These results aren't expected to be particularly good btw, so don't worry that the scores are low.

```
2025-09-29 16:30:29 INFO faster_coco_eval.core.cocoeval Evaluate annotation type *bbox*
2025-09-29 16:31:23 INFO faster_coco_eval.core.cocoeval COCOeval_opt.evaluate() finished...
2025-09-29 16:31:23 INFO faster_coco_eval.core.cocoeval DONE (t=53.93s).
2025-09-29 16:31:25 INFO faster_coco_eval.core.cocoeval Accumulating evaluation results...
2025-09-29 16:31:25 INFO faster_coco_eval.core.cocoeval COCOeval_opt.accumulate() finished...
2025-09-29 16:31:25 INFO faster_coco_eval.core.cocoeval DONE (t=0.00s).
2025-09-29 16:31:25 INFO faster_coco_eval.core.cocoeval  Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.044
2025-09-29 16:31:25 INFO faster_coco_eval.core.cocoeval  Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.166
2025-09-29 16:31:25 INFO faster_coco_eval.core.cocoeval  Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.014
2025-09-29 16:31:25 INFO faster_coco_eval.core.cocoeval  Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.005
2025-09-29 16:31:25 INFO faster_coco_eval.core.cocoeval  Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.039
2025-09-29 16:31:25 INFO faster_coco_eval.core.cocoeval  Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.073
2025-09-29 16:31:25 INFO faster_coco_eval.core.cocoeval  Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.003
2025-09-29 16:31:25 INFO faster_coco_eval.core.cocoeval  Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.019
2025-09-29 16:31:25 INFO faster_coco_eval.core.cocoeval  Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.079
2025-09-29 16:31:25 INFO faster_coco_eval.core.cocoeval  Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.003
2025-09-29 16:31:25 INFO faster_coco_eval.core.cocoeval  Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.077
2025-09-29 16:31:25 INFO faster_coco_eval.core.cocoeval  Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.112
2025-09-29 16:31:25 INFO faster_coco_eval.core.cocoeval  Average Recall     (AR) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.241
2025-09-29 16:31:25 INFO faster_coco_eval.core.cocoeval  Average Recall     (AR) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.034
```

I think the slow runtime on our cluster is partly that the CPU cores are slower (or something is). It's significantly faster on my laptop.

https://drive.google.com/file/d/1Np9qBPmTN22DPhL6EjuEDabj1KsOOYQO/view?usp=sharing

**Comment by jveitchmichaelis** [comment]
Yeah we switched to this backend and that's helped quite a lot! It's significantly faster.

**Comment by MiXaiLL76** [comment]
## pycocotools (32 minute)

```
Memory Usage Profile:
+-------------------------+----------------+----------+
| Stage                   | Memory Usage   | Diff     |
+=========================+================+==========+
| 1. Initial              | 53.48 MB       | -        |
+-------------------------+----------------+----------+
| 2. Classes imported     | 53.48 MB       | +0.00    |
+-------------------------+----------------+----------+
| 3. Ground truth loaded  | 2727.34 MB     | +2673.86 |
+-------------------------+----------------+----------+
| 4. Predictions loaded   | 4590.73 MB     | +1863.39 |
+-------------------------+----------------+----------+
| 5. Evaluator created    | 4590.73 MB     | +0.00    |
+-------------------------+----------------+----------+
| 6. Evaluation completed | 9016.98 MB     | +4426.25 |
+-------------------------+----------------+----------+
| 7. Results accumulated  | 9056.67 MB     | +39.69   |
+-------------------------+----------------+----------+
| 8. Summary generated    | 9056.67 MB     | +0.00    |
+-------------------------+----------------+----------+
```

## faster_coco_eval (1 minute) https://github.com/MiXaiLL76/faster_coco_eval/pull/62
```
Memory Usage Profile:
+-------------------------+----------------+----------+
| Stage                   | Memory Usage   | Diff     |
+=========================+================+==========+
| 1. Initial              | 54.39 MB       | -        |
+-------------------------+----------------+----------+
| 2. Classes imported     | 54.39 MB       | +0.00    |
+-------------------------+----------------+----------+
| 3. Ground truth loaded  | 2732.62 MB     | +2678.23 |
+-------------------------+----------------+----------+
| 4. Predictions loaded   | 4595.51 MB     | +1862.89 |
+-------------------------+----------------+----------+
| 5. Evaluator created    | 4595.51 MB     | +0.00    |
+-------------------------+----------------+----------+
| 6. Evaluation completed | 11398.21 MB    | +6802.70 |
+-------------------------+----------------+----------+
| 7. Results accumulated  | 11398.21 MB    | +0.00    |
+-------------------------+----------------+----------+
| 8. Summary generated    | 11398.21 MB    | +0.00    |
+-------------------------+----------------+----------+
```


## faster_coco_eval (separate_eval=False) (1 minute) https://github.com/MiXaiLL76/faster_coco_eval/pull/62
```
Memory Usage Profile:
+-------------------------+----------------+----------+
| Stage                   | Memory Usage   | Diff     |
+=========================+================+==========+
| 1. Initial              | 54.06 MB       | -        |
+-------------------------+----------------+----------+
| 2. Classes imported     | 54.06 MB       | +0.00    |
+-------------------------+----------------+----------+
| 3. Ground truth loaded  | 2732.29 MB     | +2678.23 |
+-------------------------+----------------+----------+
| 4. Predictions loaded   | 4595.19 MB     | +1862.90 |
+-------------------------+----------------+----------+
| 5. Evaluator created    | 4595.19 MB     | +0.00    |
+-------------------------+----------------+----------+
| 6. Evaluation completed | 11128.35 MB    | +6533.16 |
+-------------------------+----------------+----------+
| 7. Results accumulated  | 11244.60 MB    | +116.25  |
+-------------------------+----------------+----------+
| 8. Summary generated    | 11244.76 MB    | +0.16    |
+-------------------------+----------------+----------+
```


<img width="910" height="591" alt="Image" src="https://github.com/user-attachments/assets/b4e452ea-8a90-47ca-907e-bcc204a3af70" />

<img width="897" height="623" alt="Image" src="https://github.com/user-attachments/assets/a783b579-7e2e-4508-8f5f-f93e3fa3a23b" />

**Comment by jveitchmichaelis** [comment]
Yeah that's similar to what I see, I didn't have any issue running it on my machine but hopefully that helped stress test.

@tanishqackerman If we're able to run with a O(1M) dataset without issue, I'm not sure why you're seeing an OOM. Unless the specific geometries you have are causing a lot of memory, what is the typical number of detections per image? Other tricks are running NMS beforehand so that matching is easy (so most boxes have a single corresponding ground truth).

If you're running this through torchmetrics, make sure you remember to reset the metric at the end of your validation epoch as it will keep adding up until you run `compute()`.

**Comment by MiXaiLL76** [comment]
Overall, I don't see a problem with this. More memory, a few GB more—yes, but the speed is also 30 times faster.

These are completely incomparable values.

If there are no comments or feedback (I'll wait a week), I'll submit a PR to https://github.com/MiXaiLL76/faster_coco_eval/pull/62 and close this issue.
